transcendence-app/src/shared/db/postgres/utils.ts

```diff
@@ -36,28 +36,31 @@ export const makeQuery = async <T>(
   }
 };
 
-export function insertWithClient<T extends Record<string, any>>(
-  client: PoolClient,
-  table: table,
-  entity: T,
-) {
-  const { cols, params, values } = entityQueryMapper(entity);
-  const text = `INSERT INTO ${table}(${cols.join(', ')}) VALUES (${params.join(
-    ',',
-  )}) RETURNING *;`;
-  return client.query(text, values);
-}
-export function updateByIdWithClient<T extends Record<string, any>>(
-  client: PoolClient,
-  table: table,
-  id: string,
-  chatroom: T,
-) {
-  const { cols, values } = entityQueryMapper(chatroom);
-  const colsToUpdate = cols.map((col, i) => `${col}=$${i + 2}`).join(',');
-  const text = `UPDATE ${table} SET ${colsToUpdate} WHERE "id"=$1 RETURNING *;`;
-  return client.query(text, [id, ...values]);
-}
+export const makeTransactionalQuery = async <T>(
+  pool: PostgresPool,
+  callback: (client: PoolClient) => Promise<T | null>,
+) => {
+  let data: T | null = null;
+  try {
+    const client = await pool.connect();
+    try {
+      await client.query('BEGIN');
+      data = await callback(client);
+      await client.query('COMMIT');
+    } catch (err) {
```

Comment:
My IDE says, and I kind of agree:
> "'throw' of exception caught locally / This inspection reports any instances of JavaScript throw statements whose exceptions are always caught by containing try statements. Using throw statements as a 'goto' to change the local flow of control is likely to be confusing."

This produces a bit of code duplication, but it makes it a bit less [astonishing](url):
```
export const makeTransactionalQuery = async <T>(
  pool: PostgresPool,
  callback: (client: PoolClient) => Promise<T | null>,
) => {
  try {
    let data: T | null = null;
    const client = await pool.connect();
    try {
      await client.query('BEGIN');
      data = await callback(client);
      await client.query('COMMIT');
    } catch (err) {
      await client.query('ROLLBACK');
      if (isCriticalDatabaseError(err)) {
        PostgresLogger.error(err.message);
      }
      return null;
    } finally {
      client.release();
    }
    return data;
  } catch (err) {
    if (isCriticalDatabaseError(err)) {
      PostgresLogger.error(err.message);
    }
    return null;
  }
};
```

Reply:
Thanks :+1:. Done.